core/extension/gdextension_interface.cpp

```diff
@@ -858,84 +858,82 @@ static GDExtensionPtrUtilityFunction gdextension_variant_get_ptr_utility_functio
 }
 
 static void gdextension_string_new_with_wide_chars(GDExtensionUninitializedStringPtr r_dest, const wchar_t *p_contents) {
 	if constexpr (sizeof(wchar_t) == 2) {
-		// wchar_t is 16 bit, parse.
-		memnew_placement(r_dest, String);
-		String *dest = reinterpret_cast<String *>(r_dest);
+		// wchar_t is 16 bit (UTF-16).
+		String *dest = memnew_placement(r_dest, String);
 		dest->parse_utf16((const char16_t *)p_contents);
 	} else {
-		// wchar_t is 32 bit, copy.
-		memnew_placement(r_dest, String((const char32_t *)p_contents));
+		// wchar_t is 32 bit (UTF-32).
+		String *string = memnew_placement(r_dest, String);
+		string->parse_utf32(StrRange<char32_t>::from_c_str((const char32_t *)p_contents));
 	}
 }
 
 static void gdextension_string_new_with_latin1_chars_and_len(GDExtensionUninitializedStringPtr r_dest, const char *p_contents, GDExtensionInt p_size) {
-	memnew_placement(r_dest, String(p_contents, p_size));
+	String *dest = memnew_placement(r_dest, String);
+	dest->parse_latin1(StrRange(p_contents, _strlen_clipped(p_contents, p_size)));
 }
 
 static void gdextension_string_new_with_utf8_chars_and_len(GDExtensionUninitializedStringPtr r_dest, const char *p_contents, GDExtensionInt p_size) {
-	memnew_placement(r_dest, String);
-	String *dest = reinterpret_cast<String *>(r_dest);
+	String *dest = memnew_placement(r_dest, String);
 	dest->parse_utf8(p_contents, p_size);
 }
 
 static GDExtensionInt gdextension_string_new_with_utf8_chars_and_len2(GDExtensionUninitializedStringPtr r_dest, const char *p_contents, GDExtensionInt p_size) {
-	memnew_placement(r_dest, String);
-	String *dest = reinterpret_cast<String *>(r_dest);
+	String *dest = memnew_placement(r_dest, String);
 	return (GDExtensionInt)dest->parse_utf8(p_contents, p_size);
 }
 
 static void gdextension_string_new_with_utf16_chars_and_len(GDExtensionUninitializedStringPtr r_dest, const char16_t *p_contents, GDExtensionInt p_char_count) {
-	memnew_placement(r_dest, String);
-	String *dest = reinterpret_cast<String *>(r_dest);
+	String *dest = memnew_placement(r_dest, String);
 	dest->parse_utf16(p_contents, p_char_count);
 }
 
 static GDExtensionInt gdextension_string_new_with_utf16_chars_and_len2(GDExtensionUninitializedStringPtr r_dest, const char16_t *p_contents, GDExtensionInt p_char_count, GDExtensionBool p_default_little_endian) {
-	memnew_placement(r_dest, String);
-	String *dest = reinterpret_cast<String *>(r_dest);
+	String *dest = memnew_placement(r_dest, String);
 	return (GDExtensionInt)dest->parse_utf16(p_contents, p_char_count, p_default_little_endian);
 }
 
 static void gdextension_string_new_with_utf32_chars_and_len(GDExtensionUninitializedStringPtr r_dest, const char32_t *p_contents, GDExtensionInt p_char_count) {
-	memnew_placement(r_dest, String((const char32_t *)p_contents, p_char_count));
+	String *string = memnew_placement(r_dest, String);
+	string->parse_utf32(StrRange(p_contents, _strlen_clipped(p_contents, p_char_count)));
```

Comment:
Is `_strlen_clipped()` correct here?

That will mean that it'll clip the ending to the first `0` it encounters. However, looking at `String::copy_from_unchecked()` it seems like it would just take the length, allowing `0` to exist in the string. So, adding `_strlen_clipped()` would change behavior here.

However, I don't know if the original behavior is what we want or not, or if people depend on being able to make `String`s with `0`s in them via this function.

Reply:
> Is `_strlen_clipped()` correct here?

Unfortunately, yes! `String((const char32_t *)p_contents, p_char_count)` calls `parse_utf32(p_cstr, p_clip_to_len);` which in turn calls `parse_utf32(StrRange(p_cstr, p_cstr ? _strlen_clipped(p_cstr, p_clip_to) : 0));`. `copy_from_unchecked` is not involved.

I guess my change _does_ imply one change in that it doesn't check the incoming pointer for `null`. That should probably still be done.

Reply:
> which in turn calls `parse_utf32(StrRange(p_cstr, p_cstr ? _strlen_clipped(p_cstr, p_clip_to) : 0));`. `copy_from_unchecked` is not involved.

But doesn't `parse_utf32(const StrRange<char32_t> &p_cstr)` call to `copy_from_unchecked`?

Reply:
NUL should not exist in the Godot strings, `copy_from_unchecked` was checking for NUL it before #99816.

Reply:
Ah, ok, so then it wouldn't be a behavior change. Thanks!